jerry-core/ecma/operations/ecma-array-object.c

```diff
@@ -120,44 +159,208 @@ ecma_op_array_is_fast_array (ecma_extended_object_t *array_p) /**< ecma-array-ob
+  }
+
+  ecma_object_t *object_p = ecma_op_alloc_simple_array_object (length);
   const uint32_t aligned_length = ECMA_FAST_ARRAY_ALIGN_LENGTH (length);
-  ecma_value_t *values_p = NULL;
+  ecma_value_t *values_p;
+  values_p = (ecma_value_t *) jmem_heap_alloc_block_null_on_error (aligned_length * sizeof (ecma_value_t));
 
-  if (length != 0)
+  if (JERRY_LIKELY (values_p != NULL))
   {
-    values_p = (ecma_value_t *) jmem_heap_alloc_block_null_on_error (aligned_length * sizeof (ecma_value_t));
+    ecma_extended_object_t *ext_obj_p = (ecma_extended_object_t *) object_p;
+    ext_obj_p->u.array.u.length_prop = (uint8_t) (ext_obj_p->u.array.u.length_prop | ECMA_FAST_ARRAY_FLAG);
+    JERRY_ASSERT (object_p->u1.property_list_cp == JMEM_CP_NULL);
 
-    if (JERRY_UNLIKELY (values_p == NULL))
+    for (uint32_t i = 0; i < length; i++)
     {
-      return NULL;
+      JERRY_ASSERT (!ecma_is_value_array_hole (args_p[i]));
+      values_p[i] = ecma_copy_value_if_not_object (args_p[i]);
+    }
+
+    for (uint32_t i = length; i < aligned_length; i++)
+    {
+      values_p[i] = ECMA_VALUE_ARRAY_HOLE;
+    }
+
+    ECMA_SET_POINTER (object_p->u1.property_list_cp, values_p);
+  }
+  else
+  {
+    for (uint32_t index = 0; index < length; index++)
+    {
+      JERRY_ASSERT (!ecma_is_value_array_hole (args_p[index]));
+      ecma_builtin_helper_def_prop_by_index (object_p,
+                                             index,
+                                             args_p[index],
+                                             ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE);
     }
   }
 
-  ecma_object_t *object_p = ecma_op_new_array_object (length);
-  ecma_extended_object_t *ext_obj_p = (ecma_extended_object_t *) object_p;
+  return ecma_make_object_value (object_p);
+} /* ecma_op_new_array_object_from_buffer */
 
-  ext_obj_p->u.array.u.length_prop = (uint8_t) (ext_obj_p->u.array.u.length_prop | ECMA_FAST_ARRAY_FLAG);
-  ext_obj_p->u.array.u.hole_count += length * ECMA_FAST_ARRAY_HOLE_ONE;
+/**
+ * Allocate a new fast acces mode array object from the given property name collection
+ *
+ * Note: The given collection is freed and it's underlying buffer is reused
+ *
+ * @return ecma_value - constructed fast access mode array object
+ */
+inline ecma_value_t JERRY_ATTR_ALWAYS_INLINE
+ecma_op_new_array_object_from_prop_name_collection (ecma_collection_t *collection_p) /**< property name collection */
+{
+  return ecma_op_new_array_object_from_collection (collection_p, false);
+} /* ecma_op_new_array_object_from_prop_name_collection */
+
+/**
+ * Allocate a new fast acces mode array object from the given collection
+ *
+ * Note: The given collection is freed and it's underlying buffer is reused
+ *
+ * @return ecma_value - constructed fast access mode array object
+ */
+ecma_value_t
+ecma_op_new_array_object_from_collection (ecma_collection_t *collection_p, /**< collection to create array from */
```

Comment:
It would be good to emphasize that the original `collection` will be unavailable after calling this method.